### PR TITLE
While loops and labels

### PR DESCRIPTION
**Pull request by irvall**
Closes #75 #74

### CODE REVIEW

File: programs/main copy.py [hunk 1]
```diff
@@ -0,0 +1,16 @@
+from channel import Channel, Branch
```
Comment (fremartini) [inline programs/main copy.py:1]: remove file

File: programs/offer.py [hunk 2]
```diff
@@ -0,0 +1,23 @@
...
+        ch.send("hello!")
+        msg = ch.recv()
+
+    case Branch.RIGHT:
+        ch.send(42)
+
+
+
+""""
+Module(body=[Expr(value=Subscript(value=Name(id='Channel', ctx=Load()),
```
Comment (fremartini) [inline programs/offer.py:18]: remove comment

File: src/check.py [hunk 3]
```diff
@@ -259,12 +273,45 @@ def visit_ClassDef(self, node: ClassDef) -> None:
         env = self.pop()
 
         self.get_latest_scope().bind_nested(node.name, env)
+    """
```
Comment (fremartini) [inline src/check.py:276]: remove comment

File: src/statemachine.py [hunk 4]
```diff
@@ -180,11 +203,22 @@ def next_id():
...
                 node.accepting = True
                 return node, None
+            elif isinstance(tup, str):
+                lab = tup
+                if lab in labels:
+                    goto_trans = TGoto(lab)
+                    node.outgoing[goto_trans] = labels[lab]
+                else:
+                    forwarded_labs[lab] = node
+                    #raise SessionException(f'label {lab} not defined here')
```
Comment (fremartini) [inline src/statemachine.py:221]: remove comment?

File: src/check.py [hunk 5]
```diff
@@ -259,12 +273,45 @@ def visit_ClassDef(self, node: ClassDef) -> None:
...
+            pre_chans = nds
         for stm in node.body:
             self.visit(stm)
+        post_chans = self.get_latest_scope().get_kind(Node)
+        if pre_chans and post_chans:
+            post_chans = [chs[1] for chs in post_chans]
+            for post_chan in post_chans:
+                if post_chan.id not in self.loop_entrypoints:
+                    raise SessionException(f'loop error')
+                    #raise SessionException(f'loop error: needs to {post_chan.outgoing_action()()} {post_chan.outgoing_type()}')
```
Comment (fremartini) [inline src/check.py:313]: error message?

**Review by fremartini — CHANGES_REQUESTED** [review]

File: src/check.py [hunk 6]
```diff
@@ -260,11 +274,30 @@ def visit_ClassDef(self, node: ClassDef) -> None:
...
+                self.loop_entrypoints.add(nd.id) 
+            pre_chans = nds
         for stm in node.body:
             self.visit(stm)
+        post_chans = self.get_latest_scope().get_kind(Node)
+        if pre_chans and post_chans:
+            post_chans = [chs[1] for chs in post_chans]
+            for post_chan in post_chans:
+                if post_chan.id not in self.loop_entrypoints:
+<<<<<<< HEAD
```
Comment (fremartini) [inline src/check.py:294]: remove

**Review by fremartini — APPROVED** [review]

**Comment by irvall** [comment]
Closes #75

**Pull request merged**